### PR TITLE
Filter Command

### LINKED ISSUE #144: Add new filters to transaction list
Filters to add

1. Person
2. Description
3. Amount
4. Category
5. id

### PR DESCRIPTION
**Pull request by rollingpencil**
Resolves #144 
This PR implements the filter command.
It is able to filter by person (via phone number), date, partial description and amount.

### CODE REVIEW

File: src/main/java/spleetwaise/transaction/logic/commands/FilterCommand.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+    /**
+     * Creates an AddCommand to add the specified {@code Transaction}.
+     *
+     * @param contact The person to filter the transaction by.
+     * @param amount The amount to filter the transaction by.
+     * @param description The description to filter the transaction by.
+     * @param date The date to filter the transaction by.
+     */
+    public FilterCommand(Person contact, Amount amount, Description description, Date date) {
```
Comment (SeeYangZhi) [inline src/main/java/spleetwaise/transaction/logic/commands/FilterCommand.java:58]: Is it possible to follow how FindCommand.java is implemented?

with a `private final TxnContainsKeywordsPredicate predicate;`

**Review by SeeYangZhi — APPROVED** [review]

File: src/main/java/spleetwaise/transaction/logic/commands/FilterCommand.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+     * This method executes the filter command.
+     *
+     * @return the result of the execution.
+     */
+    @Override
+    public CommandResult execute() throws CommandException {
+        CommonModel model = CommonModel.getInstance();
+
+        Predicate<Transaction> filterPredicate = (txn) -> {
+            if (contact != null && !txn.getPerson().equals(contact)) {
```
Comment (seanlim) [inline src/main/java/spleetwaise/transaction/logic/commands/FilterCommand.java:79]: If we already did null-checks in the constructor, why do we still need to check again here? These properties are not publicly accessible right?

**Review by seanlim — APPROVED** [review]
Overall LGTM, just a few small questions

File: src/main/java/spleetwaise/transaction/logic/commands/FilterCommand.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    /**
+     * Creates an AddCommand to add the specified {@code Transaction}.
+     *
+     * @param contact The person to filter the transaction by.
+     * @param amount The amount to filter the transaction by.
+     * @param description The description to filter the transaction by.
+     * @param date The date to filter the transaction by.
+     */
+    public FilterCommand(Person contact, Amount amount, Description description, Date date) {
+        if (contact == null && amount == null && description == null && date == null) {
```
[the remark this answers is not in the document]
Comment (rollingpencil) [reply src/main/java/spleetwaise/transaction/logic/commands/FilterCommand.java:59]: In this case, I need a `requireAnyNonNull`-esk functionality, where one checks if any of the provided "filters" are non-null.

File: src/main/java/spleetwaise/transaction/logic/commands/FilterCommand.java (hunk 2 above)
Comment (rollingpencil) [reply src/main/java/spleetwaise/transaction/logic/commands/FilterCommand.java:79]: This is to check if a given person/amount/description/date is specified and, if so, whether the transaction matches the specific component.

**Pull request merged**